includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

```diff
@@ -1626,7 +1626,7 @@ private function check_attr_spec_rule_value_regex_casei( DOMElement $node, $attr
 	private function check_attr_spec_rule_valid_url( DOMElement $node, $attr_name, $attr_spec_rule ) {
 		if ( isset( $attr_spec_rule[ AMP_Rule_Spec::VALUE_URL ] ) && $node->hasAttribute( $attr_name ) ) {
 			foreach ( $this->extract_attribute_urls( $node->getAttributeNode( $attr_name ) ) as $url ) {
-				$url = urldecode( $url );
```

Comment:
I wonder why `urldecode()` is being used here. This doesn't seem to make sense. It was introduced first in 843760af15d7b66191d482b7e2b0ca8f703f6b27 as part of #983. 

However, according to https://github.com/ampproject/amp-wp/pull/983#discussion_r173029676 it seems the `urldecode()` should perhaps have been restricted to obtaining the `$host` below?

Reply:
Yeah, that's interesting.

It's hard to see how `urldecode()` would be needed to get the protocol.

Reply:
The unit tests still pass with:

```diff
diff --git a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
index d9dd93ac8..12740fc91 100644
--- a/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
+++ b/includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php
@@ -1626,7 +1626,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
        private function check_attr_spec_rule_valid_url( DOMElement $node, $attr_name, $attr_spec_rule ) {
                if ( isset( $attr_spec_rule[ AMP_Rule_Spec::VALUE_URL ] ) && $node->hasAttribute( $attr_name ) ) {
                        foreach ( $this->extract_attribute_urls( $node->getAttributeNode( $attr_name ) ) as $url ) {
-                               $url = urldecode( trim( $url ) );
+                               $url = trim( $url );
 
                                // Check whether the URL is parsable.
                                $parts = wp_parse_url( $url );
@@ -1644,7 +1644,7 @@ class AMP_Tag_And_Attribute_Sanitizer extends AMP_Base_Sanitizer {
                                }
 
                                // Check if the host contains invalid chars (hostCharIsValid: https://github.com/ampproject/amphtml/blob/af1e3a550feeafd732226202b8d1f26dcefefa18/validator/engine/parse-url.js#L62-L103).
-                               $host = wp_parse_url( $url, PHP_URL_HOST );
+                               $host = wp_parse_url( urldecode( $url ), PHP_URL_HOST );
                                if ( $host && preg_match( '/[!"#$%&\'()*+,\/:;<=>?@[\]^`{|}~\s]/', $host ) ) {
                                        return AMP_Rule_Spec::FAIL;
                                }
```

Reply:
Let's go ahead and make that change. It seems that it would make sense. The `parseUrl_` function used in the AMP validator does not do this decoding upfront. It rather only decodes the host alone: 

https://github.com/ampproject/amphtml/blob/af1e3a550feeafd732226202b8d1f26dcefefa18/validator/engine/parse-url.js#L289-L292

Reply:
Sounds good, I'll do that soon.

Reply:
Applied in https://github.com/ampproject/amp-wp/pull/4166/commits/bd4f45f17e70859845bc28d3496f8abb6642ffb3

Reply:
Applied in https://github.com/ampproject/amp-wp/pull/4166/commits/bd4f45f17e70859845bc28d3496f8abb6642ffb3